hal/tls/mbedtls/tls_mbedtls.c

```diff
@@ -945,42 +953,50 @@ TLSSocket_read(TLSSocket self, uint8_t* buf, int size)
 
                 createSecurityEvents(self->tlsConfig, ret, flags, self);
             }
+        }
 
-            return -1;
+        int reset_err = mbedtls_ssl_session_reset(&(self->ssl));
+        if (0 != reset_err) {
+            DEBUG_PRINT("TLS", "mbedtls_ssl_session_reset failed -0x%X\n", -reset_err);
         }
+
+        return ret;
     }
 
-    return ret;
+    return len;
 }
 
 int
 TLSSocket_write(TLSSocket self, uint8_t* buf, int size)
 {
-    int ret;
-    int len = size;
+    int len = 0;
 
     checkForCRLUpdate(self);
 
     if (startRenegotiationIfRequired(self) == false) {
         return -1;
     }
 
-    while ((ret = mbedtls_ssl_write(&(self->ssl), buf, len)) <= 0)
+    while (len < size)
```

Comment:
I have the feeling that this is similar to the above: you moved the logic to complete the buffer at a lower level (so you loop chunk by chunk here until whole packet is done) which IMO is already done in `CotpConnection_sendDataMessage` which will chunk at appropriate size and loop. (see around here: https://github.com/mz-automation/libiec61850/blob/2890e3f179242e90617a991557cb0b4b62f47b08/src/mms/iso_cotp/cotp.c#L182-L195 and https://github.com/mz-automation/libiec61850/blob/2890e3f179242e90617a991557cb0b4b62f47b08/src/mms/iso_cotp/cotp.c#L223-L245 and then in `CotpConnection_sendDataMessage`)

So also here I would just keep possibly the two new return codes check without modifying the state machine too much.